### DISCUSSION TITLE
A way to use `Mixins.addConfiguration()`/Alternative to?

### DISCUSSION THREAD

**Discussion by zenfyrdev** [FabricMC/fabric-api#3634] [Mod Dev Support] [answered]
### Description

In my project, I use `Mixins.addConfiguration()` to dynamically register mixin configs. The problem is that I have to rely on `IExtension` which has to be added using `MixinInternals.registerExtension()` from MixinExtras. This is very not stable and also breaks most threaded operations due to a class loading deadlock.

I'm wondering if there's a proper/better way to do this at runtime? I tried `preLaunch`, which works fine, until some mod loads a Minecraft class. `IMixinConfigPlugin` also has a classloading deadlock and using `addConfiguration` here crashes the game with `ConcurrentModificationException`.

I also thought about adding configs at compile time, but that comes with its own set of problems. 

Is it just impossible to do cleanly?

**Comment by LlamaLad7** [comment, marked as answer]
Calling `Mixins.addConfiguration` from `preLaunch` is the correct approach. Any mod that loads minecraft classes in `preLaunch` is broken.
On another note you should absolutely not touch anything internal from MixinExtras, at all. It can change at any time.

**Reply by zenfyrdev** [reply]
Hmm, guess I'll have to retry preLaunch. Thank you.

> It can change at any time.

I know, it's just that the method was already there when I was testing extensions. 😅
